add: Бронеплиты

vovakons opened the pull request:
## Что этот ПР делает
Добавление бронеплит в игру. Теперь броня от баллистики и лазеров в игре есть только через бронеплиты (*за очень редким исключением*), у всего остального защита от пуль и лазеров удалена.
Бронеплиты защищают от баллистики/лазера согласно уровню бронепробития (*смотрите таблицы*), при поглощении урона плита ломается и теряет защитные параметры (*полностью сломанная плита уже не защищит ни от чего, сломанная наполовину - защитит вполовину от своей эффективности*).

Бронеплиты можно крафтить (*самодельные*) и ремонтировать определенным ресурсом.
Можно купить в сб вендоре и вендоре оружейки. Для антагов в аплинк добавил покупку бронеплит комплектом. Также можно заказать плиты в карго.

[Таблицы для баланса](https://docs.google.com/spreadsheets/d/11zL202kRN3uBJecr65o3_-sHkskBggEzRTCLIktJINg/edit?usp=sharing)

## Почему это хорошо для игры
Вечная защита в игре это не очень балансно, дальнее оружие в игре сейчас сильно уступает мили (*особенно после моих обнов с разбросом и точностью*). Данный пр сделает броню ломаемой и дальнее оружие станет более эффективным. А уровни бронепробития позволят не опираться на тупо цифры, а на классы боеприпасов и классы защиты.

## Демонстрация изменений
Плиты:
<img width="586" height="368" alt="image" src="https://github.com/user-attachments/assets/0e0b5e6a-f9ef-4f05-ba4d-f76a1df18cd1" />
Описание через шифтклик:
<img width="676" height="307" alt="image" src="https://github.com/user-attachments/assets/4723e98f-6198-456e-a0ed-e11e83c55003" />
атак логи показывающие падение показателя брони при повреждении:
<img width="672" height="357" alt="image" src="https://github.com/user-attachments/assets/5035c121-38a6-4d16-af87-4627a3b225ea" />

## Тестирование
Протестировано на локалке. Но требуется тестмерж чтобы найти нормальный баланс.

vovakons committed:
basic armor plate item

vovakons committed:
new defines file for armors

vovakons committed:
armor plate basics

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
basic plates

vovakons committed:
suits armor plate setup

vovakons committed:
plate attach and remove procs

Перемещение проков для добавления и удаления бронеплит на броню.
Определены зоны покрытия бронеплит.

vovakons committed:
basic armor plate algorith

Базовый алгоритм просчета бронеплит при попадании снарядов.
Исправлен examine для бронеплит.
Исправлен прок get_armor_efficient, теперь возвращает значение в
диапазоне от 0 до 1.
Временно добавлен armour_penetration_level (уровень бронепробития пули) для прожектайлов.

vovakons committed:
set bullets armor penetration level

Установлены уровни бронепробития для всех пуль (баллистики).

PlayerUnknown14 commented:
Тарков апдейт???

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
laser beams penetration setup

Установка значения бронепробиваемости для лазеров всех видов.
Добавлен новый тип бронепробития для лазеров - не пробивающий. Нужен для
лазертагов.

vovakons committed:
armor plates in security vending

Добавлены легкие плиты в продажу для торгомата сб.
Добавлены средние плиты (кроме керамики) в раздатчик боеприпасов сб за
кредиты. Туда потому что он находится в оружейке.

vovakons committed:
armor start plates logic

Некоторые виды брони изначально спавнятся с плитой.
Добавлено поле для определения стартовой плиты у брони.

vovakons committed:
helmet armor plate

Встроенные бронеплиты на каски.
Добавление бронеплит для большинства шлемов и сютов.

vovakons committed:
hurdsuit bugfix

vovakons committed:
armor plates for suits and helmets

Завершение установки плит для брони и касок.
Вроде все нашел.

vovakons committed:
head armor plate logic

Добавлена обработка бронеплиты внутри шлемов.

vovakons committed:
fix helmet armor plate bug

vovakons committed:
head plates fix

Фикс бронеплит внутри шлемов.

littleboobs commented on code/__DEFINES/armor_penetration.dm:
```diff
@@ -0,0 +1,71 @@
+/// Heavy rifles: 5.56x45mm, 4.6x30mm AP
+#define BALLISTIC_PENETRATION_HEAVY_RIFLE	6
+/// High calibers: .50COMP и .50
+#define BALLISTIC_PENETRATION_HIGH_CAL		7
```
Всё хорошо, но не надо табуляции перед значениями ставить, пожалуйста!!

littleboobs commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,368 @@
+			return 34 * plate.get_armor_efficient()
+		if(-3)
+			return 20 * plate.get_armor_efficient()
+	return 0
```
Дефайнизировать бы это.

littleboobs commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,368 @@
+/proc/calculate_armor_plate_damage_mod_for_laser(obj/item/armor_plate/plate, penetration_level)
+	var/penetration_delta = plate.laser_class - penetration_level
+	switch(penetration_delta)
+		if(0)
+			return 1
+		if(1)
+			return 0.75
+		if(2)
+			return 0.5
+		if(3 to 100)
+			return 0.25
+		if(-1)
+			return 1.25
+		if(-2)
+			return 1.5
+		if(-100 to -3)
+			return 2
+	return 0
```
итс э мейджик :)

vovakons replied:
> итс э мейджик :)

ага, выглядит пока всрато, подумаю еще

littleboobs commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,368 @@
+			return 2
+	return 0
+
+// MARK: Handmade armor plates
+/obj/item/armor_plate/handmade_steel
+	name = "handmade steel armor plate"
+	icon_state = "steelplate_light" //TODO need icon
+	plate_slot = ARMOR_PLATE_SLOT_HANDMADE
+	ballistic_class = BALLISTIC_ARMOR_CLASS_I
+	laser_class = LASER_ARMOR_CLASS_NONE
+	armor_protection_integrity = 50
+	armor_max_integrity = 50
+	body_parts_covered = UPPER_TORSO
```
всем им понадобятся рунеймсы)

vovakons replied:
😭

littleboobs commented on code/modules/mob/living/simple_animal/friendly/snake.dm:
```diff
@@ -164,7 +164,7 @@
 	regenerate_icons()
 
-/mob/living/simple_animal/hostile/retaliate/poison/snake/rouge/getarmor(def_zone, attack_flag)
+/mob/living/simple_animal/hostile/retaliate/poison/snake/rouge/getarmor(def_zone, attack_flag, penetration_level = 1, damage = 0)
```
тут полагаю тоже должны быть дефайны, явно не бул

vovakons committed:
remove bullet and laser defence from clothing

vovakons committed:
remove all defence from bullets and laser from clothing

Удаление оставшейся брони от пуль и лазеров, замена на бронеплиты.

vovakons committed:
armor plate refactor and helmet plates

Унификация бронеплит для всей одежды.
Добавление новых типов бронеплит для шлемов с покрытием головы.

vovakons committed:
more armor plate unification

Перенос логики крепления плит с сьюта и шлема на одежду.
Логика создания бронеплит на броне сделана через armor_plate переменную.

vovakons committed:
plate examine info fix

vovakons committed:
examine text fix

vovakons committed:
space suit jetpack remove and armor plate remove bug fix

vovakons committed:
fix attach and detach armor plates from wearing armors

vovakons committed:
detail info in clothing and armor plate

vovakons committed:
cargo crates for armor plates

vovakons committed:
basic penetration define

vovakons committed:
craft handmate armor plates

vovakons committed:
armor plates ballistic class rename

vovakons committed:
armor plates inhands

Добавлены спрайты на руках для бронеплит.
Добавлен спрайт для элитной бронеплиты.
Изменен баланс крепления бронеплит, больше различных плит можно вешать
на различную броню.

vovakons committed:
armor plates integrity balance

Баланс прочности бронеплит - плиты стали прочнее на 50 единиц.

PlayerUnknown14 committed:
Локализация

PlayerUnknown14 committed:
Merge branch 'armor-rework' of https://github.com/vovakons/Paradise into armor-rework

vovakons committed:
armor plate penetration balance refactor

Датумизация проков баланса для бронепробития плит.

vovakons committed:
Merge branch 'armor-rework' of github.com:vovakons/Paradise into armor-rework

vovakons committed:
remove plates with hands

vovakons committed:
armor plate repair timing

PlayerUnknown14 committed:
Доработка локализации

PlayerUnknown14 committed:
Merge branches 'armor-rework' and 'armor-rework' of https://github.com/vovakons/Paradise into armor-rework

vovakons committed:
armor slowdown feature

vovakons committed:
Merge branch 'armor-rework' of github.com:vovakons/Paradise into armor-rework

PlayerUnknown14 committed:
звуки велкро

PlayerUnknown14 committed:
Merge branch 'master220' into armor-rework

PlayerUnknown14 committed:
Звуки в 22 грц

PlayerUnknown14 committed:
Merge branch 'armor-rework' of https://github.com/vovakons/Paradise into armor-rework

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
disable fast load

vovakons committed:
handmade plates sprites

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
repair and integrite text fix

vovakons commented:
Готов к тестмержу

UnderwaterNeike commented:
Готов к фаст мержу, фаст деплою. 🐈💨

dageavtobusnick committed:
Merge branch 'master220' into armor-rework

vovakons committed:
fix names spaces to tabs

dageavtobusnick committed:
Merge branch 'master220' into armor-rework

littleboobs committed:
Merge branch 'master220' into armor-rework

littleboobs commented on code/__DEFINES/combat.dm:
```diff
@@ -39,22 +39,22 @@
+#define CANPARALYSE 	(1<<4)
+#define CANPUSH	 		(1<<5)
+#define PASSEMOTES 		(1<<6)      //Mob has a cortical borer or holders inside of it that need to see emotes.
+#define IGNORESLOWDOWN 	(1<<7)
```
Верни. Ознакомься – https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#tabs-not-spaces

littleboobs commented on code/datums/supplypacks.dm:
```diff
@@ -874,6 +874,96 @@ GLOBAL_LIST_INIT(all_supply_groups, list(SUPPLY_EMERGENCY,SUPPLY_SECURITY,SUPPLY
+	containername = "ящик тяжёлых бронеплит"
+
+
+
```
Контролируй колво пустого места, которое создаёшь, пожалуйста.
```suggestion
	containername = "ящик тяжёлых бронеплит"

```

littleboobs commented on code/modules/crafting/tailoring.dm:
```diff
@@ -496,3 +496,22 @@
 	result = /obj/item/clothing/glasses/monocle
 	reqs = list(/obj/item/clothing/glasses/thermal/monothermal = 1)
+
+
```
```suggestion

```

littleboobs commented on code/modules/mob/living/living_defense.dm:
```diff
@@ -10,31 +10,31 @@
 */
-/mob/living/proc/run_armor_check(def_zone, attack_flag = MELEE, absorb_text, soften_text, armour_penetration, penetrated_text)
-	var/armor = getarmor(def_zone, attack_flag)
+/mob/living/proc/run_armor_check(def_zone, attack_flag = MELEE, absorb_text, soften_text, armour_penetration, penetrated_text, armour_penetration_level = 1, damage = 0)
```
дефайн на пен лвл

littleboobs commented on code/modules/mob/living/living_defense.dm:
```diff
@@ -180,17 +180,17 @@
 				return
-		M.occupant_message(span_danger("[pluralize_ru(M.occupant.gender,"Ты","Вы")] ударяе[pluralize_ru(M.occupant.gender,"шь","те")] [declent_ru(ACCUSATIVE)]."))
-		visible_message(span_danger("[capitalize(M.declent_ru(NOMINATIVE))] ударя[pluralize_ru(M.gender,"ет","ют")] [declent_ru(ACCUSATIVE)]!"), span_userdanger("[capitalize(M.declent_ru(NOMINATIVE))] ударя[pluralize_ru(M.gender,"ет","ют")] [pluralize_ru(src.gender,"тебя","вас")]!"))
+		M.occupant_message(span_danger("Вы ударяеете [declent_ru(ACCUSATIVE)]."))
```
```suggestion
		M.occupant_message(span_danger("Вы ударяете [declent_ru(ACCUSATIVE)]."))
```

vovakons replied:
клава беспороводная иногда разряжается и пролагивает 😅

littleboobs commented on code/modules/projectiles/projectile.dm:
```diff
@@ -63,6 +63,7 @@
 	var/flag = BULLET
 	///How much armor this projectile pierces.
 	var/armour_penetration = 0
+	var/armour_penetration_level = 1
```
дефайн и комм

littleboobs commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,596 @@
+// Armor plates.
+// Can be attached to armor. Give bonus armor for laser and bullet
+
+
+//MARK: Basic armor plate
+
+/obj/item/armor_plate
```
```suggestion
// Armor plates.
// Can be attached to armor. Give bonus armor for laser and bullet

//MARK: Basic armor plate
/obj/item/armor_plate
```

littleboobs commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,596 @@
+/obj/item/armor_plate/proc/get_examine_text()
+	. = span_notice("Установлен[genderize_ru(gender, "", "а", "о", "ы")] <b>[declent_ru(NOMINATIVE)]</b>. ")
+	. += span_notice("\n- [get_armor_text()]")
+	. += span_notice("\n- [get_integrity_text()]")
```
```suggestion
	. += span_notice("\n– [get_armor_text()]")
	. += span_notice("\n– [get_integrity_text()]")
```

littleboobs reviewed (changes requested):
Мелкие исправление, не относящиеся к локализации в основном.

dageavtobusnick commented on code/modules/clothing/clothing.dm:
```diff
@@ -56,6 +56,19 @@
+
+/obj/item/clothing/Initialize(mapload)
+	. = ..()
+	if(ispath(armor_plate))
```
ранний ретурн

dageavtobusnick commented on code/modules/clothing/upgrade_modules/plates.dm:
```diff
@@ -0,0 +1,596 @@
+
+/// Take armor damage proc
+/obj/item/armor_plate/proc/take_armor_damage(damage_amount)
+	armor_integrity = max(armor_integrity - damage_amount, 0)
```
Уверен, что тебе не подойдет obj_integrity и нужно городить этот велосипед?

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
review fixes

vovakons committed:
hotfix

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
linter fix 1

LoneAsket commented:
В идеале добавить бы еще иконку разлома брони справа сверху как состояния, либо отображение на самой броне визуал разлома.
ИМХО довольно неудобно каждый раз шифт-кликать по себе, чтобы узнать состояние этой самой бронеплиты.

PlayerUnknown14 committed:
Merge branch 'master220' into armor-rework

PlayerUnknown14 commented:
> В идеале добавить бы еще иконку разлома брони справа сверху как состояния, либо отображение на самой броне визуал разлома. ИМХО довольно неудобно каждый раз шифт-кликать по себе, чтобы узнать состояние этой самой бронеплиты.

Слишком казуально будет. Не соревновательный шутан же, функции осмотра хватит с головой.

dageavtobusnick commented:
> > В идеале добавить бы еще иконку разлома брони справа сверху как состояния, либо отображение на самой броне визуал разлома. ИМХО довольно неудобно каждый раз шифт-кликать по себе, чтобы узнать состояние этой самой бронеплиты.
> 
> Слишком казуально будет. Не соревновательный шутан же, функции осмотра хватит с головой.

Удобство != казуальность. Хорошая сложность ставит перед игроком препятствие, которое он должен преодолеть, проявив навык. А просмотр состояния брони через shift + click это просто дрочево.

PlayerUnknown14 commented:
Ватевер.

vovakons committed:
destroyed armor plate overlay icon

Логика показа оверлей иконки если бронеплита сломана.

vovakons committed:
armor plate overlay improvements

Теперь стутсный оверлей на броне при ношении с плитой показывает тип
бронеплиты и его состояние:
- красный: сломан
- синий: поврежден и не дает 100% защиту
- зеленый: цел и дает 100% своей защиты

vovakons commented:
перезапустите пж проверку бота

vovakons committed:
empty commit for wakeup bot

vovakons committed:
Merge branch 'master220' into armor-rework

vovakons committed:
remove duplicate variables in armor plates

vovakons committed:
remove duplicate variables

dageavtobusnick committed:
Merge branch 'master220' into armor-rework

vovakons committed:
armor plates in uplink

Добавлены киты с бронеплитами в аплинк.
5 тк - легкие.
10 тк - сердние.
18 тк - тяжелые.

vovakons committed:
Merge branch 'armor-rework' of github.com:vovakons/Paradise into armor-rework

PlayerUnknown14 committed:
Merge branch 'master220' into armor-rework

vovakons committed:
remove handmade plate from fire suits

vovakons committed:
Merge branch 'armor-rework' of github.com:vovakons/Paradise into armor-rework

vovakons committed:
fire suit with handmade ablative plate but can not remove

vovakons committed:
ghostface armor with ceramic plate

vovakons committed:
add kevlar plate recipe

Добавлен рецепт крафта кевларовой брони из дюраткани.

vovakons committed:
remove armor plate from fake magical hat

vovakons committed:
armor plate repair coefficient increase

vovakons committed:
increase armors integrity

vovakons committed:
armor plate kit box sprite

vovakons committed:
armor plate destroy logic and 7.62x54 ammo penetration balance

vovakons committed:
remove armored berets

vovakons committed:
reflector special plate

vovakons committed:
integrated armor fixes

Встроенные бронеплиты теперь при шифт клике называются встроенным
бронеэлементом.
Фикс извлечения сломанных бронеплит.
Добавлен ремонт встроенных бронеплит.

vovakons committed:
riot armor balance

риот броня теперь имеет кевларовую плиту со старта и позвоялет нацепить
только плиты самодельного класса.

vovakons committed:
makeshift armor fix

vovakons committed:
remove fake wizard hat armor

vovakons committed:
disable sound when damage armor plate

vovakons committed:
fix code

vovakons committed:
decrease bone helbet armor

vovakons committed:
fix balance table

vovakons committed:
fixes 2

vovakons committed:
armor indicator on ground fix

vovakons committed:
armor plate integrity text with examine

vovakons committed:
add armor plate in human examine

Добавлен показ плиты при осмотре куклы.

vovakons committed:
armor slowdown fix

vovakons committed:
hardsuits fix

vovakons committed:
space suit slowdown fix

vovakons committed:
sound effect woth impact and destroy plate

vovakons committed:
space ninja suit fix

vovakons commented:
Казуалы победили

UnderwaterNeike commented:
Плаксы победили.

SoulFN commented:
Гады

Hexagel commented:
Наныли

NuSanya commented:
> Наныли

<img width="412" height="150" alt="image" src="https://github.com/user-attachments/assets/50d8c157-961c-4916-8df5-069983f42f55" />

pilygun commented:
На одни бронеплиты смотрели двое...

tsoivecen199015 commented:
НЕЕЕЕЕЕЕТ!!!
